### ISSUE TITLE
Unable to install templates on Ubuntu 18.04 after upgrading to 2.1.302

### ISSUE THREAD

**Issue by nguerrera** [dotnet/corefx#31259]
_From @NatElkins on July 21, 2018 22:23_

## Steps to reproduce
Try installing any dotnet template with dotnet version 2.1.302 on Ubuntu 18.04.  For example, `dotnet new -i "Amazon.Lambda.Templates::*"`.  But it happens with any of them.

## Expected  behavior
The templates should install without any issues.

## Actual behavior
You'll get an error like:
```
nathaniel@nelknet:~/Projects$ dotnet new -i "Amazon.Lambda.Templates::*"
  Restoring packages for /home/nathaniel/.templateengine/dotnetcli/v2.1.302/scratch/restore.csproj...
/usr/share/dotnet/sdk/2.1.302/NuGet.targets(114,5): error : Unable to load the service index for source https://api.nuget.org/v3/index.json. [/home/nathaniel/.templateengine/dotnetcli/v2.1.302/scratch/restore.csproj]
/usr/share/dotnet/sdk/2.1.302/NuGet.targets(114,5): error :   The type initializer for 'System.Net.Http.HttpConnectionPool' threw an exception. [/home/nathaniel/.templateengine/dotnetcli/v2.1.302/scratch/restore.csproj]
/usr/share/dotnet/sdk/2.1.302/NuGet.targets(114,5): error :   Version's parameters must be greater than or equal to zero. [/home/nathaniel/.templateengine/dotnetcli/v2.1.302/scratch/restore.csproj]
/usr/share/dotnet/sdk/2.1.302/NuGet.targets(114,5): error :   Parameter name: revision [/home/nathaniel/.templateengine/dotnetcli/v2.1.302/scratch/restore.csproj]

```
and the templates will not be installed.

## Environment data
`dotnet --info` output:

```
nathaniel@nelknet:~/Projects$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.302
 Commit:    9048955601

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64
 Base Path:   /usr/share/dotnet/sdk/2.1.302/

Host (useful for support):
  Version: 2.1.2
  Commit:  811c3ce6c0

.NET Core SDKs installed:
  2.0.2 [/usr/share/dotnet/sdk]
  2.1.4 [/usr/share/dotnet/sdk]
  2.1.302 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

_Copied from original issue: dotnet/cli#9713_

**Comment by nguerrera** [comment]
_From @NatElkins on July 22, 2018 18:18_

This persists even after uninstalling all dotnet SDKs, removing ~/.nuget, ~/.templateengine, and ~/.dotnet, and also removing /usr/share/dotnet.  However, I have another Ubuntu 18.04 machine, and it doesn't appear to have this problem.  Any ideas on how I can track down the cause would be deeply appreciated.

**Comment by nguerrera** [comment]
_From @NatElkins on July 22, 2018 19:6_

With verbosity set to detailed, I'm seeing this:
```
     1>Target "Restore" in file "/usr/share/dotnet/sdk/2.1.302/NuGet.targets" from project "/home/nathaniel/Projects/debug-nuget-issue/fs/fs.fsproj" (entry point):
       Task "RemoveDuplicates"
       Done executing task "RemoveDuplicates".
       Using "RestoreTask" task from assembly "/usr/share/dotnet/sdk/2.1.302/NuGet.Build.Tasks.dll".
       Task "RestoreTask"
         (in) RestoreGraphItems Count '5'
         (in) RestoreDisableParallel 'False'
         (in) RestoreNoCache 'False'
         (in) RestoreIgnoreFailedSources 'False'
         (in) RestoreRecursive 'True'
         (in) RestoreForce 'False'
         (in) HideWarningsAndErrors 'False'
         Running restore with 4 concurrent jobs.
         Reading project file /home/nathaniel/Projects/debug-nuget-issue/fs/fs.fsproj.
         Restoring packages for /home/nathaniel/Projects/debug-nuget-issue/fs/fs.fsproj...
         Restoring packages for .NETCoreApp,Version=v2.1...
     1>/usr/share/dotnet/sdk/2.1.302/NuGet.targets(114,5): error : Unable to load the service index for source https://api.nuget.org/v3/index.json. [/home/nathaniel/Projects/debug-nuget-issue/fs/fs.fsproj]
/usr/share/dotnet/sdk/2.1.302/NuGet.targets(114,5): error :   The type initializer for 'System.Net.Http.HttpConnectionPool' threw an exception. [/home/nathaniel/Projects/debug-nuget-issue/fs/fs.fsproj]
/usr/share/dotnet/sdk/2.1.302/NuGet.targets(114,5): error :   Version's parameters must be greater than or equal to zero. [/home/nathaniel/Projects/debug-nuget-issue/fs/fs.fsproj]
/usr/share/dotnet/sdk/2.1.302/NuGet.targets(114,5): error :   Parameter name: revision [/home/nathaniel/Projects/debug-nuget-issue/fs/fs.fsproj]
         NuGet.Protocol.Core.Types.FatalProtocolException: Unable to load the service index for source https://api.nuget.org/v3/index.json. ---> System.TypeInitializationException: The type initializer for 'System.Net.Http.HttpConnectionPool' threw an exception. ---> System.ArgumentOutOfRangeException: Version's parameters must be greater than or equal to zero.
         Parameter name: revision
            at System.Version..ctor(Int32 major, Int32 minor, Int32 build, Int32 revision)
            at System.Environment.<>c.<.cctor>b__79_0()
            at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
            at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
            at System.Lazy`1.CreateValue()
            at System.Environment.get_OSVersion()
            at System.Net.Http.HttpConnectionPool.GetIsWindows7Or2008R2()
            at System.Net.Http.HttpConnectionPool..cctor()
            --- End of inner exception stack trace ---
            at System.Net.Http.HttpConnectionPool.ConstructSslOptions(HttpConnectionPoolManager poolManager, String sslHostName)
            at System.Net.Http.HttpConnectionPool..ctor(HttpConnectionPoolManager poolManager, HttpConnectionKind kind, String host, Int32 port, String sslHostName, Uri proxyUri, Int32 maxConnections)
            at System.Net.Http.HttpConnectionPoolManager.SendAsyncCore(HttpRequestMessage request, Uri proxyUri, Boolean doRequestAuth, Boolean isProxyConnect, CancellationToken cancellationToken)
            at System.Net.Http.HttpConnectionPoolManager.SendAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
            at System.Net.Http.HttpAuthenticatedConnectionHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
            at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
            at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
            at NuGet.Protocol.ServerWarningLogHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
            at NuGet.Protocol.HttpSourceAuthenticationHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
            at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
            at NuGet.Protocol.TimeoutUtility.StartWithTimeout[T](Func`2 getTask, TimeSpan timeout, String timeoutMessage, CancellationToken token)
            at NuGet.Protocol.HttpRetryHandler.SendAsync(HttpRetryHandlerRequest request, ILogger log, CancellationToken cancellationToken)
            at NuGet.Protocol.HttpSource.GetThrottledResponse(Func`1 requestFactory, TimeSpan requestTimeout, TimeSpan downloadTimeout, Int32 maxTries, Guid sessionId, ILogger log, CancellationToken cancellationToken)
            at NuGet.Protocol.HttpSource.<>c__DisplayClass14_0`1.<<GetAsync>b__0>d.MoveNext()
         --- End of stack trace from previous location where exception was thrown ---
            at NuGet.Common.ConcurrencyUtilities.ExecuteWithFileLockedAsync[T](String filePath, Func`2 action, CancellationToken token)
            at NuGet.Protocol.HttpSource.GetAsync[T](HttpSourceCachedRequest request, Func`2 processAsync, ILogger log, CancellationToken token)
            at NuGet.Protocol.ServiceIndexResourceV3Provider.GetServiceIndexResourceV3(SourceRepository source, DateTime utcNow, ILogger log, CancellationToken token)
            --- End of inner exception stack trace ---
            at NuGet.Protocol.ServiceIndexResourceV3Provider.GetServiceIndexResourceV3(SourceRepository source, DateTime utcNow, ILogger log, CancellationToken token)
            at NuGet.Protocol.ServiceIndexResourceV3Provider.TryCreate(SourceRepository source, CancellationToken token)
            at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T](CancellationToken token)
            at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T]()
            at NuGet.Protocol.HttpFileSystemBasedFindPackageByIdResourceProvider.TryCreate(SourceRepository sourceRepository, CancellationToken token)
            at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T](CancellationToken token)
            at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T]()
            at NuGet.Commands.SourceRepositoryDependencyProvider.EnsureResource()
            at NuGet.Commands.SourceRepositoryDependencyProvider.FindLibraryCoreAsync(LibraryRange libraryRange, NuGetFramework targetFramework, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
            at NuGet.Commands.SourceRepositoryDependencyProvider.<>c__DisplayClass18_0.<<FindLibraryAsync>b__0>d.MoveNext()
         --- End of stack trace from previous location where exception was thrown ---
            at NuGet.Commands.SourceRepositoryDependencyProvider.FindLibraryAsync(LibraryRange libraryRange, NuGetFramework targetFramework, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
            at NuGet.DependencyResolver.ResolverUtility.<>c__DisplayClass7_1.<<FindLibraryFromSourcesAsync>b__0>d.MoveNext()
         --- End of stack trace from previous location where exception was thrown ---
            at NuGet.DependencyResolver.ResolverUtility.FindLibraryFromSourcesAsync(LibraryRange libraryRange, IEnumerable`1 providers, Func`2 action)
            at NuGet.DependencyResolver.ResolverUtility.FindLibraryByVersionAsync(LibraryRange libraryRange, NuGetFramework framework, IEnumerable`1 providers, SourceCacheContext cacheContext, ILogger logger, CancellationToken token)
            at NuGet.DependencyResolver.ResolverUtility.FindLibraryMatchAsync(LibraryRange libraryRange, NuGetFramework framework, GraphEdge`1 outerEdge, IEnumerable`1 remoteProviders, IEnumerable`1 localProviders, IEnumerable`1 projectProviders, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
            at NuGet.DependencyResolver.ResolverUtility.FindLibraryEntryAsync(LibraryRange libraryRange, NuGetFramework framework, GraphEdge`1 outerEdge, RemoteWalkContext context, CancellationToken cancellationToken)
            at NuGet.DependencyResolver.RemoteDependencyWalker.CreateGraphNode(LibraryRange libraryRange, NuGetFramework framework, String runtimeName, RuntimeGraph runtimeGraph, Func`2 predicate, GraphEdge`1 outerEdge)
            at NuGet.DependencyResolver.RemoteDependencyWalker.CreateGraphNode(LibraryRange libraryRange, NuGetFramework framework, String runtimeName, RuntimeGraph runtimeGraph, Func`2 predicate, GraphEdge`1 outerEdge)
            at NuGet.Commands.ProjectRestoreCommand.WalkDependenciesAsync(LibraryRange projectRange, NuGetFramework framework, String runtimeIdentifier, RuntimeGraph runtimeGraph, RemoteDependencyWalker walker, RemoteWalkContext context, CancellationToken token)
            at NuGet.Commands.ProjectRestoreCommand.TryRestoreAsync(LibraryRange projectRange, IEnumerable`1 frameworkRuntimePairs, NuGetv3LocalRepository userPackageFolder, IReadOnlyList`1 fallbackPackageFolders, RemoteDependencyWalker remoteWalker, RemoteWalkContext context, Boolean forceRuntimeGraphCreation, CancellationToken token)
            at NuGet.Commands.RestoreCommand.ExecuteRestoreAsync(NuGetv3LocalRepository userPackageFolder, IReadOnlyList`1 fallbackPackageFolders, RemoteWalkContext context, CancellationToken token)
            at NuGet.Commands.RestoreCommand.ExecuteAsync(CancellationToken token)
            at NuGet.Commands.RestoreRunner.ExecuteAsync(RestoreSummaryRequest summaryRequest, CancellationToken token)
            at NuGet.Commands.RestoreRunner.ExecuteAndCommitAsync(RestoreSummaryRequest summaryRequest, CancellationToken token)
            at NuGet.Commands.RestoreRunner.CompleteTaskAsync(List`1 restoreTasks)
            at NuGet.Commands.RestoreRunner.RunAsync(IEnumerable`1 restoreRequests, RestoreArgs restoreContext, CancellationToken token)
            at NuGet.Commands.RestoreRunner.RunAsync(RestoreArgs restoreContext, CancellationToken token)
            at NuGet.Build.Tasks.RestoreTask.ExecuteAsync(ILogger log)
       Done executing task "RestoreTask" -- FAILED.
     1>Done building target "Restore" in project "fs.fsproj" -- FAILED.
     1>Done Building Project "/home/nathaniel/Projects/debug-nuget-issue/fs/fs.fsproj" (Restore target(s)) -- FAILED.
```
Looks like it is failing on this line: https://github.com/dotnet/corefx/blob/a3021f143b762006a7e19ad11785ca0c3c454933/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L22

**Comment by nguerrera** [comment]
@karelz @Petermarcu The stack here has Environment.OSVersion causing an ArgumentOutOfRangeException and preventing HttpConnectionPool from getting through its static constructor. I think this should be considered for 2.1.x servicing.

**Comment by nguerrera** [comment]
@NatElkins Than you for the detailed report.

**Comment by nguerrera** [comment]
@NatElkins Can you provide the output of `uname -a` for the machine where this is not working for you and for the machine where it is working. The stack here involves handling data returned from `uname` system call.

**Comment by nguerrera** [comment]
If I'm reading the code right, the only way I'm seeing a negative revision happening is an integer overflow here: https://github.com/dotnet/corefx/blob/6ed61acebe3214fcf79b4274f2bb9b55c0604a4d/src/System.Runtime.Extensions/src/System/Environment.Unix.cs#L355

Which I think could happen if there's a string of digits in `utsname.release` that would parse to greater than int.MaxValue.

**Comment by nguerrera** [comment]
cc @stephentoub

**Comment by NatElkins** [comment]
@nguerrera Thanks for looking into this so quickly!  From the non-working machine, `uname -a` returns:
```
Linux nelknet 4.15.0-24201807041620-generic #0+mediatree+hauppauge-Ubuntu SMP Thu Jul 5 01:38:17 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
Some kind of patched kernel was required in order to support a TV tuner that I use with the computer.  Instructions for that are here: http://hauppauge.com/pages/support/support_linux.html

For the working machine, `uname -a` returns:
```
Linux nelknet-carbon 4.15.0-23-generic #25-Ubuntu SMP Wed May 23 18:02:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by nguerrera** [comment]
Bingo: revision is `24201807041620`

**Comment by NatElkins** [comment]
@nguerrera So is the best workaround right now to change my kernel?

**Comment by danmosemsft** [comment]
@NatElkins yes, either change the kernel to one that does not have this large number, or revert back to your earlier version of .NET Core. What did you upgrade from? Assuming you did. I would expect any 2.1.x version to hit this. The particular callstack did not exist in 2.0 but I am a little surprised no other codepath used `Environment.OSVersion` but in a quick look they seem to only be for Windows versions - indeed this one is for Windows, but gets the OSVersion on Unix as well.

**Comment by danmosemsft** [comment]
We should fix this though, and probably service for it so you can actually use this kernel.

**Comment by danmosemsft** [comment]
Another option is potentially to patch your .NET Core with one you have built yourself to avoid the problem. This is an advanced scenario. https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md

**Comment by nguerrera** [comment]
That's the only one I can see.

**Comment by NatElkins** [comment]
Damn @danmosemsft, you're an animal! :)

**Comment by NatElkins** [comment]
Looking forward to testing this when it's released.

**Comment by danmosemsft** [comment]
@NatElkins you're welcome but note it will probably be 2 or more likely 3 months before it goes out in servicing (if approved which I think it will be)

**Comment by danmosemsft** [comment]
reactivating for 2.1.x

**Comment by danmosemsft** [comment]
#### Description
On some Linux kernels, `dotnet new` will not work (and probably many other things)

Details: On Linux we parse `uname` for version components. However `uname` can contain almost anything. If it includes and unusually large number we may overflow causing an exception soon after when we try to represent it in a Version object. This is triggered by simply getting `Environment.OSVersion`. The first component to do this is currently `HttpConnectionPool`.
		 
#### Customer Impact
Customer using kernel customized by Haupppage cannot use .NET Core. May be other kernels broken. No workaround short of modifying and recompiling your kernel.
		 
#### Regression?
From .NET Framework. May have occurred in previous .NET Core, although this callstack is relatively new it would probably break later.
		 
#### Risk
Very low - catch overflow exception.

**Comment by danmosemsft** [comment]
@NatElkins is it possible for you to build master locally and verify that it fixes your problem?

**Comment by danmosemsft** [comment]
Approved by big shiproom

**Comment by vivmishra** [comment]
Approved for 2.1.4

**Comment by joshfree** [comment]
Moving to 2.1.5 as the PR is incomplete and has outstanding feedback to address still.

**Comment by joshfree** [comment]
@tarekgh could you load balance this from @danmosemsft ?

**Comment by NatElkins** [comment]
:(

**Comment by danmosemsft** [comment]
The PR for release/2.1  https://github.com/dotnet/corefx/pull/31332 is now updated and ready to go. Thanks @tarekgh

**Comment by danmosemsft** [comment]
Moving label to PR per new process

**Comment by tarekgh** [comment]
@danmosemsft I assigned this back to you. let me know if need any help from my side. Thanks.

**Comment by danmosemsft** [comment]
@tarekgh nope, thanks for fixing it 😊

**Comment by karelz** [comment]
It was fixed in release/2.1 branch in PR #31332. It will be part of 2.1.5 release.
